### DISCUSSION TITLE
Why does my GPDCtorch run slower than GPDC?

### DISCUSSION THREAD

**Discussion by bxt14** [jakobrunge/tigramite#155] [Q&A]
I was trying to use GPDCtorch to speed up the nonlinear causal discovery calculations. GPDCtorch can compute the results. However, to my surprise, it runs even slower than GPDC. 

Here's my data generation process.

`import numpy as np
from matplotlib import pyplot as plt
T = 200
np.random.seed(1)
data = np.random.randn(T, 4)
for t in range(1, T):
    data[t, 1] += 0.4*data[t-2, 0]**2
    data[t, 2] += 0.2*data[t-4, 1]**2
    data[t, 3] += 0.5*data[t-3, 2]**2 - 0.6*data[t-5, 1]
var_names=['A','B','C','D']

from tigramite import data_processing as pp
from tigramite import plotting as tp
from tigramite.pcmci import PCMCI
from tigramite.independence_tests import ParCorr, GPDC, CMIknn, CMIsymb, GPDCtorch

dataframe = pp.DataFrame(data, var_names=var_names)
tp.plot_timeseries(dataframe)
plt.show()`

Then I use GPDC for causal discovery.
`
gpdc = GPDC(significance='analytic', gp_params=None)
pcmci_gpdc = PCMCI(
    dataframe=dataframe, 
    cond_ind_test=gpdc,
    verbosity=0)

results = pcmci_gpdc.run_pcmci(tau_max=5, pc_alpha=0.1)
pcmci_gpdc.print_significant_links(
        p_matrix = results['p_matrix'], 
        val_matrix = results['val_matrix'],
        alpha_level = 0.01)
`
![1](https://user-images.githubusercontent.com/75287774/137090297-4184ef36-bc30-4eb7-ba84-be5fa989da48.png)
The calculation took 21.6 seconds.

Then I used GPDCtorch instead.
`
gpdctorch = GPDCtorch(significance='analytic')
pcmci_gpdc_torch = PCMCI(
    dataframe=dataframe, 
    cond_ind_test=gpdctorch,
    verbosity=0)

results = pcmci_gpdc_torch.run_pcmci(tau_max=5, pc_alpha=0.1)
pcmci_gpdc_torch.print_significant_links(
        p_matrix = results['p_matrix'], 
        val_matrix = results['val_matrix'],
        alpha_level = 0.01)
`
![2](https://user-images.githubusercontent.com/75287774/137093657-9a295a77-2211-432d-9e18-ba4def2958fb.png)
The calculation took 187.8s.

So I'm wondering what's wrong with my use of GPDCtorch?

**Comment by sagarsimha** [comment]
Hi. Can you give me details of the system you are running it on - the interface, the system configuration etc? I see Number of devices = 2. Are you running it on 2 GPUs?

**Reply by bxt14** [reply]
The computations are implemented on a computing server with Intel Xeon Gold 5118 CPU and NVIDIA TITAN RTX GPU.

![01CPU](https://user-images.githubusercontent.com/75287774/138219759-6fb2729d-6d11-46fa-bd1c-a4499687a676.png)
![02GPU](https://user-images.githubusercontent.com/75287774/138219775-22f52a88-ae46-4e7e-b743-22e6e29b78e3.png)

The programming interface is as follows
![03Program](https://user-images.githubusercontent.com/75287774/138219801-af0ea237-dd07-4cbf-86b7-937ef32a7e11.png)

**Comment by jakobrunge** [comment]
We've had quite some updates, please open an issue if it still persists.
